src/psyclone/tests/psyir/frontend/fparser2_data_statement_test.py

```diff
@@ -0,0 +1,128 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+'''Tests Fortran data statements in the fparser2 PSyIR front-end. Data
+statements are represented as an UnsupportedFortranType. The tests
+here cover a few variations of the data statement, in case that at
+some stage we need more information.
+'''
+
+import pytest
+from fparser.common.readfortran import FortranStringReader
+from fparser.two.Fortran2003 import Specification_Part
+from psyclone.psyir.frontend.fparser2 import Fparser2Reader
+from psyclone.psyir.nodes import Routine
+from psyclone.psyir.symbols import UnsupportedFortranType
+
+
+@pytest.mark.usefixtures("f2008_parser")
+def test_simple_data_statement():
+    ''' Test that data statements are correctly captured. '''
+
+    # Create a dummy test routine
+    routine = Routine("test_routine")
+    symtab = routine.symbol_table
+    processor = Fparser2Reader()
+
+    # Test with a single namelist
+    reader = FortranStringReader('''
+        data a /1/''')
+    fparser2spec = Specification_Part(reader)
+    processor.process_declarations(routine, fparser2spec.content, [])
+    # There is a data_stmt in the symbol table:
+    data = symtab.lookup("_PSYCLONE_INTERNAL_DATA_STMT")
+    assert isinstance(data.datatype, UnsupportedFortranType)
+    assert data.datatype.declaration == "DATA a / 1 /"
+
+
+@pytest.mark.usefixtures("f2008_parser")
+def test_multiple_data_statement():
+    ''' Test that data statements are correctly captured. '''
+
+    # Create a dummy test routine
+    routine = Routine("test_routine")
+    symtab = routine.symbol_table
+    processor = Fparser2Reader()
+
+    # Test with a single namelist
+    reader = FortranStringReader('''
+        data a /1/, b/2/, c/3/''')
+    fparser2spec = Specification_Part(reader)
+    processor.process_declarations(routine, fparser2spec.content, [])
+    # There is a data_stmt in the symbol table:
+    data = symtab.lookup("_PSYCLONE_INTERNAL_DATA_STMT")
+    assert isinstance(data.datatype, UnsupportedFortranType)
+    assert data.datatype.declaration == "DATA a / 1 /, b / 2 /, c / 3 /"
+
+
+@pytest.mark.usefixtures("f2008_parser")
+def test_data_statement_implicit_loop():
+    ''' Test that z data statement with an implicit loops is correctly
+    captured. '''
+
+    # Create a dummy test routine
+    routine = Routine("test_routine")
+    symtab = routine.symbol_table
+    processor = Fparser2Reader()
+
+    # Test with a single namelist
+    reader = FortranStringReader(
+        "DATA (es(ies),ies=    0, 5) / 0.966483e-02,0.966483e-02,"
```

Comment:
Oh yes. This makes it very clear why we only have partial support :-) (I was starting to think that it would be easy to convert into a canonical form. Although, having said that, it shouldn't be *too* difficult to add suitable statements. Not something for this PR though.)

Reply:
You are aware implied do-loops can be nested?  :)

Reply:
"walk in the park" :-)